### PR TITLE
add tlog server and client

### LINKED ISSUE #49: tlog server implementation in go
This would allow us to share some code between the tlog server and the nbdserver
like the pooling of the ardb's, failover mechanism, storagecluster configuration
...

the embedded ledisdb (on which Glen is working) for inmemory storage during development,...

and publishing the hashses to the rivine blockchain could import the required libraries directly from rivine without the need for reimplementation

since that is implemented in go as well

### PR DESCRIPTION
**Pull request by iwanbk**
Closes #49 

Some of the sub task at #49 are moved to new issue because i don't think anymore that we must do it.

### CODE REVIEW

**Comment by GlenDC** [comment]
Are you sure this is already ready to merge into master? I thought we were still doing work on this? I'm certainly still working on some feature.

Also I think before this can be merged, we should clean up the commit history of this branch, as it's now pretty messy.

**Comment by iwanbk** [comment]
```
Are you sure this is already ready to merge into master?
```
Yes, it won't break anything. And make the integration to be much easier because the ndbserver master still changing.

```
 I thought we were still doing work on this
```
Yes, we will always working on this.

```
Also I think before this can be merged, we should clean up the commit history of this branch, as it's now pretty messy.
```
right

**Comment by GlenDC** [comment]
@iwanbk ok that does make sense.

I'm ok with this being merged into master then, as long as you clean up the commit history a bit, as it's now pretty messy as I said (especially all those commits at the beginning).

**Comment by GlenDC** [comment]
BTW, I do think that it would be cool if issue #92 could be resolved before merging in this PR. I opened a PR for it.

**Comment by iwanbk** [comment]
```
(especially all those commits at the beginning).
```
yes, it is very messy.
As it supposed to be cleaned before sending the PR, i never had any problem this way before but got many many conflicts in this branch, that involved code not written by me.
It seems because this branch was modified not only by me and involve rebase/merge from master.

 The command i use to squashed it
```
git rebase -i ced30d1b4c0c3b749603ec3fdbb78324c2d4bbd9
```
i got the hash from 
```
$ git merge-base tlogserver master
ced30d1b4c0c3b749603ec3fdbb78324c2d4bbd9
```

If someone really knows how to handle it, please do it or tell me. 

If not, we can merge this to master

**Comment by GlenDC** [comment]
Yeah I don't think we can (and should) all squash it into one commit. I think it's better to squash the commits that clearly belong together, but as this is a very big PR, it's ok that there are a couple of different commits. Going to work on merging this later today.

**Comment by iwanbk** [comment]
yep, i didn't try to make it a single commit.
I only tried to squash the commits from the first day (IIRC April 13th).

**Comment by GlenDC** [comment]
Cleaned up the tlogserver branch as much as I ought reasonable. I left alone all master-to-tlogserver merge commits as to not mess up.

Will do a final self-review after lunch, and will merge it into master afterwards.

A backup (of the old tlogserver) branch is pushed to https://github.com/g8os/blockstor/tree/tlogserver-backup, that one will remain online for the next week or so, in case we notice something wrong because of the tlogserver-history-rewriting.

Should be fine though, as there weren't any unexpected merge conflicts. Only the ones that were already there.

File: tlog/install_dependencies.sh [hunk 1]
```diff
@@ -0,0 +1,26 @@
+#!/bin/bash
+
+BUILD_DIR=/tmp/build_tlog
+rm -rf $BUILD_DIR
+mkdir $BUILD_DIR
+
+
+sudo apt-get install build-essential autoconf automake nasm yasm libsnappy-dev
```
Comment (GlenDC) [inline tlog/install_dependencies.sh]: Only works for Debian for now. This script should also work for MacOS and other required platforms once we know for sure that we know for sure that we want to keep the CGO code.

File: tlog/tlogclient/README.md [hunk 2]
```diff
@@ -0,0 +1,17 @@
+# TLOG Client
+
+- TLOG Client library is a client lib that works on GIG BLOCKSTOR (NBD Server)
+- TLOG Client communicates with TLOG Server via dpdk stack.
+- Communication between client and server use binary capnp stream (no RPC).
+
+## Data send from the client to server:
+```
+Block:
```
Comment (GlenDC) [inline tlog/tlogclient/README.md]: Actualize these docs, as they seem to be out of date and incorrect.

File: tlog/tlogclient/client.go [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+}
+
+// Send sends the transaction tlog to server.
+// It returns error in these cases:
+// - broken network
+// - failed to send all tlog
+// in case of errors, client is not in valid state,
+// shouldn't be used anymore
+func (c *Client) Send(vdiskID string, seq uint64,
+	lba, timestamp uint64, data []byte) error {
```
Comment (GlenDC) [inline tlog/tlogclient/client.go]: should move to Line 156. This is the only line in the entire codebase that we break up like this, so better be consistent and combine line 156 and 157 into one.

File: tlog/tlogclient/decoder/decoder.go [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	// get the real len
+	var realLen uint64
+	if err := binary.Read(bytes.NewBuffer(merged), binary.LittleEndian, &realLen); err != nil {
+		return nil, err
+	}
+	merged = merged[8 : 8+realLen] /* 8 is the size of uint64 */
```
Comment (GlenDC) [inline tlog/tlogclient/decoder/decoder.go]: 8 should move to a constant and calculated using sizeof

**Pull request merged**